content/docs/using-pulumi/debugging/_index.md

```diff
@@ -0,0 +1,295 @@
+
+Then, attach your debugger to the process with the `node` runtime. For example, in VS Code, you can define the following `launch.json` file:
+
+```json
+{
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "name": "Node.js: Attach",
+            "type": "node",
+            "request": "attach",
+            "continueOnAttach": true,
+            "skipFiles": [
+                "${workspaceFolder}/node_modules/**/*.js",
+                "${workspaceFolder}/lib/**/*.js",
+                "<node_internals>/**/*.js"
+            ]
+        }
+    ]
+}
+```
+
+navigate to the `Run and Debug` menu on the left pane, and then click the Start Debugging (F5) button with the `Node.js: Attach` configuration.
+
+Depending on your setup, you may need to attach multiple times until the right node process is found. For example, you may be running Node.js policy packs that will hit the debugger as well.
+
+Now you can step through the program and inspect variables. Once completed, resume the execution and let the Pulumi CLI complete.
+
+## Debugging Python programs
+
+This guide uses [`debugpy`](https://github.com/microsoft/debugpy) to debug Pulumi programs.
+
+You need to install the `debugpy` package to your Pulumi program by adding it to `requirements.txt`:
+
+```txt
+pulumi>=3.0.0,<4.0.0
+debugpy>=1.0.0
+```
+
+Then, prepend your Pulumi program with the code to wait for a debugger to attach:
+
+```py
+import debugpy
+
+debugpy.listen(("localhost", 5678))
+debugpy.wait_for_client()
+
+# ... Your normal Pulumi code follows here
+```
+
+Run `pulumi up` the usual way:
+
+```bash
+pulumi up
+Previewing update (dev)
+
+ Type                 Name  Plan   Info
+ pulumi:pulumi:Stack  dev          0.00s - Note: Debugging will proceed.
+```
+
+Set a breakpoint in your Pulumi code.
+
+Navigate to your IDE and attach the debugger to the process with the `python` runtime. For example, in VS Code, you can define the following `launch.json` file:
+
+```json
+{
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "name": "Python: Local Attach",
+            "type": "debugpy",
+            "request": "attach",
+            "connect": {
+                "host": "localhost",
+                "port": 5678
+            },
+            "pathMappings": [
+                {
+                    "localRoot": "${workspaceFolder}",
+                    "remoteRoot": "."
+                }
+            ],
+            "justMyCode": true
+        }
+    ]
+}
+```
+
+navigate to the `Run and Debug` menu on the left pane, and then click the Start Debugging (F5) button with the `Python: Local Attach` configuration.
+
+Now you can step through the program and inspect variables. Once completed, resume the execution and let the Pulumi CLI complete.
+
+## Debugging .NET programs
+
+Prepend your Pulumi program with the code to wait for a debugger to attach:
+
+```cs
+using System.Diagnostics;
+
+return await Pulumi.Deployment.RunAsync(async () =>
```

Comment:
Might be worth mentioning the `async` bit is necessary here, as without it, you'll get a very noisy compile error.

Reply:
Added a comment to this effect